### PR TITLE
Schedule task with derivative account

### PR DESCRIPTION
**Pull request by imstar15**
Related OAK-blockchain PR:
https://github.com/OAK-Foundation/OAK-blockchain/pull/499

Dependencies:
https://github.com/OAK-Foundation/oak.js/pull/143
https://github.com/OAK-Foundation/oak.js/pull/144

**Task  was scheduled**

image

**Task  was executed**

image

Log:

```
1. Transfer TUR to the derivative account on turing-local: 
status.type Ready
status.type InBlock
status.type Finalized

2. One-time proxy setup on moonbase-local

a) Add a proxy for Alice If there is none setup on moonbase-local

oakChainData.paraId:  2114
Add a proxy of moonbase-local (paraId:1000) and proxyType: Any on Turing ...
 Proxy address: 0xfde409e557153fceffd60db438375b2247df3418

status.type Ready
status.type InBlock
status.type Finalized

b) Topping up the proxy account on moonbase-local with DEV ...

status.type Ready
status.type InBlock
status.type Finalized

User Alice turing-local address: 6AwtFW6sYcQ8RcuAJeXdDKuFtUVXj4xW57ghjYQ5xyciT1yd, moonbase-local address: 0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac

3. Execute an XCM from moonbase-local to turing-local ...

a). Create a payload to store in Turing’s task ...

b). Send extrinsic from moonbase-local to turing-local to schedule task. Listen to TaskScheduled event on turing-local chain ...
Listen XCMP task events
Send extrinsic from moonbase-local to schedule task. extrinsic: 0x21060301010009210103010100092101dc6a337f000000000000000000000000e1023c010004000000000000000003010100a10f03010100092103010200a10f040300141354778000000000000000000000c9012601f24ff3a9cf04c71dbc94d0b566f7a27b94566cac01581501000000000000000000000000000000000000000000000000000000000000a72f549a1a12b9b49f30a7f3aeb1f4e96389c5d8000000000000000000000000000000000000000000000000000000000000000010d09de08a0003404ac76c5a15010003401462a85a8603000100031d4a5a434a38010001071d72c531014a38010000
status.type Ready
        automationTime:TaskScheduled:: (phase={"applyExtrinsic":0})
                        AccountId32: 6AubHEyP5NhszkySgTAznkkQpiBjnBRkh2EidUQFv8k7kaPK
                        Bytes: 0x31362d302d31
                        Option<AccountId32>: 
Found the event and retrieved TaskId, 16-0-1

Keep Listening automationTime.TaskExecuted until 2024-01-09 10:00:52(1704765652) to verify that the task(taskId: 16-0-1) will be successfully triggered ...
status.type InBlock
        automationTime:TaskExecuted:: (phase=Initialization)
                        AccountId32: 6AubHEyP5NhszkySgTAznkkQpiBjnBRkh2EidUQFv8k7kaPK
                        Bytes: 0x31362d302d31
XcmpMessageSent event:  {
  phase: 'Initialization',
  event: {
    method: 'XcmpMessageSent',
    section: 'xcmpQueue',
    index: '0x2804',
    data: {
      messageHash: '0x03c026ccf147d5277559da361e7d6ddc28709f002fcf240cfa35ee816ca0a51a'
    }
  },
  topics: []
}
messageHash:  0x03c026ccf147d5277559da361e7d6ddc28709f002fcf240cfa35ee816ca0a51a
status.type Finalized

4. Listen xcmpQueue.Success event with messageHash(0x03c026ccf147d5277559da361e7d6ddc28709f002fcf240cfa35ee816ca0a51a) and find ethereum.Executed event on Parachain...
        xcmpQueue:Success:: (phase={"applyExtrinsic":1})
                        Option<[u8;32]>: 0x03c026ccf147d5277559da361e7d6ddc28709f002fcf240cfa35ee816ca0a51a
                        {"refTime":"Compact<u64>","proofSize":"Compact<u64>"}: {"refTime":813047000,"proofSize":28373}
ethereum.Executed event:  {"from":"0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac","to":"0xa72f549a1a12b9b49f30a7f3aeb1f4e96389c5d8","transactionHash":"0xe8d1b7bf8ae46af6b1a31f10d50f2e39f0cab13905c0361770c2a64b50f606a8","exitReason":{"Succeed":"Stopped"},"extraData":""}
Reached the end of main() ...
```

### CODE REVIEW

**Review by chrisli30 — COMMENTED** [review]
There’s a catch to this flow.

On our Automation Hub, the fee on Turing Network is paid through TUR, but not DEV correct?

If yes, do we still need the step of bridge DEV to the proxy account on Turing? Could you modify this file to remove this step and use TUR instead?
```
1.  Reserve transfer DEV to the proxy account on turing-local: 
paraTokenIdOnOak:  5
minBalanceOnOak:  500000000000000000
paraTokenbalanceOnOak.free:  0
Transfer DEV from Moonbase to Turing
Transfer from moonbase-local, extrinsic: 
```

**Comment by imstar15** [comment]
> There’s a catch to this flow.
> 
> On our Automation Hub, the fee on Turing Network is paid through TUR, but not DEV correct?
> 
> If yes, do we still need the step of bridge DEV to the proxy account on Turing? Could you modify this file to remove this step and use TUR instead?
> 
> ```
> 1.  Reserve transfer DEV to the proxy account on turing-local: 
> paraTokenIdOnOak:  5
> minBalanceOnOak:  500000000000000000
> paraTokenbalanceOnOak.free:  0
> Transfer DEV from Moonbase to Turing
> Transfer from moonbase-local, extrinsic: 
> ```

Yes, Solved.

File: src/moonbeam/common.js [hunk 1]
```diff
@@ -135,14 +111,16 @@ export const scheduleTask = async ({
...
         destinationChainAdapter: moonbeamAdapter,
         taskPayloadExtrinsic,
-        scheduleFeeLocation: moonbeamDefaultAsset.location,
-        executionFeeLocation: moonbeamDefaultAsset.location,
-        scheduleAs: u8aToHex(keyringPair.addressRaw),
+        scheduleFeeLocation: oakAsset.location,
+        executionFeeLocation: moonbeamAsset.location,
         keyringPair: moonbeamKeyringPair,
+        invoker: moonbeamAdapter,
+        invokeFeeLocation: oakAsset.location,
```
Comment (imstar15) [inline src/moonbeam/common.js:123]: Add the 'invokeFeeLocation' parameter to specify the asset for the XCM message fee when invoking the scheduled task.

Add the 'invokeFeeLocation' parameter to specify the invoker chain.

File: src/moonbeam/common.js [hunk 2]
```diff
@@ -151,7 +129,7 @@ export const scheduleTask = async ({
     console.log(`Found the event and retrieved TaskId, ${taskId}`);
 
     const executionTime = scheduleActionType === ScheduleActionType.executeOnTheHour
-        ? nextExecutionTime : moment().valueOf() / 1000;
+        ? nextExecutionTime : (Math.floor(moment().valueOf() / 1000) + 60);
     const timeout = calculateTimeout(nextExecutionTime);
```
Comment (chrisli30) [inline src/moonbeam/common.js:133]: `calculateTimeout(nextExecutionTime)` didn’t use the calculated executionTime.

File: src/moonbeam/common.js (hunk 1 above)
Comment (chrisli30) [reply src/moonbeam/common.js:123]: let’s use `caller` instead of `invoker`, and `callerXcmFeeLocation` to replace `invokeFeeLocation`

**Review by chrisli30 — APPROVED** [review]
The PR looks good. I left two comments.

**Review by chrisli30 — APPROVED** [review]
The code looks good. Let’s upgrade oak.js libraries first and test this PR with npm packages.

File: src/moonbeam/common.js [hunk 3]
```diff
@@ -139,50 +114,34 @@ export const scheduleTask = async ({
...
-
-    const executionTime = scheduleActionType === ScheduleActionType.executeOnTheHour
-        ? nextExecutionTime : moment().valueOf() / 1000;
-    const timeout = calculateTimeout(nextExecutionTime);
-
-    console.log(`\n4. Keep Listening events on ${parachainName} until ${moment(executionTime * 1000).format('YYYY-MM-DD HH:mm:ss')}(${executionTime}) to verify that the task(taskId: ${taskId}) will be successfully executed ...`);
 
-    const listenEventsResult = await listenEvents(oakApi, 'automationTime', 'TaskTriggered', { taskId }, timeout);
+    const executionTime = scheduleActionType === ScheduleActionType.executeImmediately ? (Math.floor(moment() / 1000) + 60) : nextExecutionTime;
+    const [, { taskId, messageHash }] = await waitPromises([sendExtrinsicPromise, listenXcmpTaskEvents(oakApi, executionTime)]);
```
Comment (imstar15) [inline src/moonbeam/common.js:125]: Merge the listening logic for TaskScheduled and TaskTriggered events, addressing an issue where waiting for extrinsic finalization was blocking the listening of TaskTriggered, thus causing the TaskTriggered event to be inaccessible.

Comment (chrisli30) [reply src/moonbeam/common.js:125]: Okay, nice.

File: src/moonbeam/common.js (hunk 1 above)
Comment (Lalsha40) [reply src/moonbeam/common.js]: I don't know my pr code
I need pr code Oak

**Pull request merged**